### PR TITLE
LanguageSensor handle new DictReadOnlyWrapper type

### PR DESCRIPTION
**Pull request by BoltzmannBrain**
The LanguageSensor region assigns values to the `outputs` array, and thus fails following the changes in [nupic.core #931](https://github.com/numenta/nupic.core/pull/931). This was caught by the Bamboo builds of Imbu ([log here](https://ci.numenta.com/download/IMBU-IMBU-JOB1/build_logs/IMBU-IMBU-JOB1-186.log)).

### CODE REVIEW

**Comment by mrcslws** [comment]
That change only breaks already-broken code. These values weren't actually being written to output. So I think this is a success story?

I'm surprised this `.update` approach is doing anything... looking now.

**Comment by mrcslws** [comment]
I'd expect `update` to fail on a `DictReadOnlyWrapper`.

I think these lines should be:

~~~python
outputs["sourceOut"][:len(data["_token"])] = data["_token"]
self.encoder.encodeIntoArray(data["_token"], outputs["dataOut"])
outputs["encodingOut"][:len(outputs["dataOut"])] = outputs["dataOut"]
~~~

It's untested. It requires them all to be declared in the spec. But it will be something like that. You have to put the values into the existing array, rather than point the dict items to a different array. I made that change because it's really easy to do it "wrong" in Python, and it was really bad that it failed silently.

**Comment by BoltzmannBrain** [comment]
@mrcslws I was using `update` (which you're right, does fail) b/c the `"sourceOut"` and `"encodingOut"` aren't declared in the spec; [`"sourceOut"` b/c it's a string](https://github.com/numenta/nupic.research/blob/master/htmresearch/regions/LanguageSensor.py#L180-L181). Do you know a better way to do this?

**Comment by mrcslws** [comment]
I haven't had to put a string in output yet, but I think one solution is to output to a Byte array.

Another is to not use the outputs. Currently it's a no-op, adding it to the dict. The only reason this accomplishes anything is because the code saves off the result via `self._outputValues = copy.deepcopy(outputs)`. So a quick fix would be to stop saving that copy of the outputs, and just save specific fields to `self` that won't fit into Network API. Then use `region.getOutputData` to get all network outputs, and get the strings directly from the `self` (maybe with a getter).

**Comment by BoltzmannBrain** [comment]
`"sourceOut"` is useful as extra info (to see what the specific token is when we're debugging) but not necessary data for the network, so limiting it to `self._outputValues` suffices. Setting `outputs["encodingOut"]` was a no-op.

@subutai please review. This PR fixes LanguageSensor issues with assigning values to the output array. Imbu tests and NLP model API tests passing confirm these changes.

**Comment by subutai** [comment]
@BoltzmannBrain How come this shows changes from another PR (renaming resetSequence) that was already merged in.  Do you need to merge master into this PR again?

**Comment by BoltzmannBrain** [comment]
@subutai the `resetSequence` renamings here should've been included in the previous PR but were overlooked.

File: htmresearch/regions/LanguageSensor.py [hunk 1]
```diff
@@ -162,37 +162,33 @@ def compute(self, inputs, outputs):
...
 
     if self.verbosity > 0:
       print "LanguageSensor outputs:"
       print "SeqID: ", outputs["sequenceIdOut"]
       print "Categories out: ", outputs['categoryOut']
-      print "dataOut: ",outputs["dataOut"].nonzero()[0]
+      print "dataOut: ", outputs["dataOut"].nonzero()[0]
 
-    self._outputValues = copy.deepcopy(outputs)
+    self._outputValues = {field: value for field, value in outputs.iteritems()}
```
Comment (subutai) [inline htmresearch/regions/LanguageSensor.py]: I'm not sure this is the right implementation. You need to copy the values. I don't think this list comprehension is making a copy of the values. 

What was wrong with the previous deepcopy line?

File: htmresearch/regions/LanguageSensor.py [hunk 2]
```diff
@@ -162,37 +162,33 @@ def compute(self, inputs, outputs):
...
-    # explicitly b/c PyRegion.getSpec() won't take an output field w/ type str.
+    # Copy important data input fields over to outputs dict.
     outputs["resetOut"][0] = data["_reset"]
     outputs["sequenceIdOut"][0] = data["_sequenceId"]
-    outputs["sourceOut"] = data["_token"]
-    self.populateCategoriesOut(data["_category"], outputs['categoryOut'])
-    outputs["encodingOut"] = self.encoder.encodeIntoArray(
-      data["_token"], outputs["dataOut"])
+    self.populateCategoriesOut(data["_category"], outputs["categoryOut"])
+    self.encoder.encodeIntoArray(data["_token"], outputs["dataOut"])
```
Comment (subutai) [inline htmresearch/regions/LanguageSensor.py]: This line looks correct. The previous assignment to outputs["encodingOut"] was actually incorrect.

File: htmresearch/regions/LanguageSensor.py (hunk 1 above)
Comment (BoltzmannBrain) [inline htmresearch/regions/LanguageSensor.py]: Using deepcopy makes `self._outputValues` a [`DictReadOnlyWrapper` type](https://github.com/numenta/nupic.core/blob/master/bindings/py/nupic/bindings/regions/PyRegion.py#L32), which by definition does not allow us to add a `"sourceOut"` field for the token string.

Comment (subutai) [inline htmresearch/regions/LanguageSensor.py]: OK, that makes sense. Then my guess is you need to do {field: copy.deepcopy(value) ...}  The arrays in the outputs dict correspond to fixed memory buffers in C++ that are setup by the Network API, so if you want to store them you have to do a copy.

Comment (BoltzmannBrain) [inline htmresearch/regions/LanguageSensor.py]: Hmm could you explain this more? I tested this and it looks like the deep copy is not necessary.

Comment (BoltzmannBrain) [inline htmresearch/regions/LanguageSensor.py]: Nvmd, @mrcslws clarified 🙈

**Pull request merged**